jnigen/java/src/main/java/com/github/dart_lang/jnigen/apisummarizer/disasm/AsmTypeUsageSignatureVisitor.java

```diff
@@ -0,0 +1,99 @@
+// Copyright (c) 2022, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+package com.github.dart_lang.jnigen.apisummarizer.disasm;
+
+import com.github.dart_lang.jnigen.apisummarizer.elements.TypeUsage;
+import java.util.ArrayList;
+import org.objectweb.asm.signature.SignatureVisitor;
+
+public class AsmTypeUsageSignatureVisitor extends SignatureVisitor {
+  private final TypeUsage typeUsage;
+
+  public AsmTypeUsageSignatureVisitor(TypeUsage typeUsage) {
+    super(AsmConstants.API);
+    this.typeUsage = typeUsage;
+  }
+
+  @Override
+  public void visitBaseType(char descriptor) {
+    typeUsage.kind = TypeUsage.Kind.PRIMITIVE;
+    var name = "";
+    switch (descriptor) {
+      case 'Z':
+        name = "boolean";
+        break;
+      case 'B':
+        name = "byte";
+        break;
+      case 'C':
+        name = "char";
+        break;
+      case 'D':
+        name = "double";
+        break;
+      case 'F':
+        name = "float";
+        break;
+      case 'I':
+        name = "int";
+        break;
+      case 'J':
+        name = "long";
+        break;
+      case 'L':
+        name = "object";
+        break;
+      case 'S':
+        name = "short";
+        break;
+      case 'V':
+        name = "void";
+        break;
+    }
+    typeUsage.shorthand = name;
+    typeUsage.type = new TypeUsage.PrimitiveType(name);
```

Comment:
Nit: should we not re-create `PrimitiveType`s for the same `name` every time?

(Feel free to ignore if it needlessly complicates the code.)

Reply:
I plan to completely change the structure to be more tree like https://github.com/dart-lang/jnigen/issues/157. These will all change then. Parsing that structure would require writing our own parser but it's worth it and we won't have to depend of JsonSerializable.